src/object/cli_obj.c

```diff
@@ -4923,33 +4924,39 @@ obj_dup_sgls_free(struct obj_auxi_args *obj_auxi)
 
 				if (skip_sgl_iov(false, iov)) {
 					j++;
+					sg_nr_out++;
 					continue;
 				}
 
 				/* Direct copy if entry wasn't modified */
 				if (!isset_range((uint8_t *)ctx->merged_bitmaps[i], j, j)) {
 					*iov = *iov_dup;
-					D_ASSERT(dup_buf_size == 0);
+					D_ASSERT(dup_data_len == 0);
 					j++;
 					sg_nr_out++;
 					dup_sg_idx++;
 					continue;
 				}
-				if (dup_buf_size == 0) {
+				if (dup_data_len == 0) {
+					dup_data_len = iov_dup->iov_len;
 					dup_buf_size = iov_dup->iov_buf_len;
 					dup_buf      = (char *)iov_dup->iov_buf;
 				}
 
+				/* Update iov_len for short read */
+				if (dup_data_len < iov->iov_len)
+					iov->iov_len = dup_data_len;
 				/* Copy data from duplicate buffer to original buffer */
-				D_ASSERT(dup_buf_size >= iov->iov_buf_len);
-				memcpy((char *)iov->iov_buf, dup_buf, iov->iov_buf_len);
-				dup_buf_size -= iov->iov_buf_len;
-				dup_buf += iov->iov_buf_len;
+				D_ASSERT(dup_buf_size >= iov->iov_len);
+				memcpy((char *)iov->iov_buf, dup_buf, iov->iov_len);
```

Comment:
Some confused here, for read case, the output `lov_len` depend on the read result, right? Why not copy as much data as possible as long as the output buffer is large enough? something like that:
```
if (iov->iov_buf_len > dup_data_len)
   iov->iov_len = dup_data_len;
else
   iov->iov_len = iov->iov_buf_len;
memcpy((char *)iov->iov_buf, dup_buf, iov->iov_len);
```

On the other hand, it the application requires only copy specified length to the given buffer which real size is large than the required length (`iov->iov_len < iov->iov_buf_len`), then how can we guarantee that all the fetched data can be filled into the given SGL buffer? I may miss something.